### PR TITLE
week6 content

### PR DESCRIPTION
**Pull request by taljmars**

### CODE REVIEW

File: Scala2-Functional-Programming/Week6/forcomp/src/main/scala/forcomp/Anagrams.scala [hunk 1]
```diff
@@ -33,10 +35,10 @@ object Anagrams extends AnagramsInterface {
    *
    *  Note: you must use `groupBy` to implement this method!
    */
-  def wordOccurrences(w: Word): Occurrences = ???
+  def wordOccurrences(w: Word): Occurrences = w.groupBy(c=>c.toLower).map(x => (x._1, x._2.length)).toList.sorted
```
Comment (maryum375) [inline Scala2-Functional-Programming/Week6/forcomp/src/main/scala/forcomp/Anagrams.scala:38]: Generally, we don't like seeing `_1` and `_2` in the code. it makes reading more difficult.
Specially here, there is a simpler solution.
Try using `mapValues` func

File: Scala2-Functional-Programming/Week6/forcomp/src/main/scala/forcomp/Anagrams.scala [hunk 2]
```diff
@@ -53,10 +55,11 @@ object Anagrams extends AnagramsInterface {
    *    List(('a', 1), ('e', 1), ('t', 1)) -> Seq("ate", "eat", "tea")
    *
    */
-  lazy val dictionaryByOccurrences: Map[Occurrences, List[Word]] = ???
+  lazy val dictionaryByOccurrences: Map[Occurrences, List[Word]] = dictionary.groupBy(w => wordOccurrences(w)).withDefaultValue(List())
+
 
   /** Returns all the anagrams of a given word. */
-  def wordAnagrams(word: Word): List[Word] = ???
+  def wordAnagrams(word: Word): List[Word] = dictionaryByOccurrences(wordOccurrences(word))
```
Comment (maryum375) [inline Scala2-Functional-Programming/Week6/forcomp/src/main/scala/forcomp/Anagrams.scala:62]: Error handling?

File: Scala2-Functional-Programming/Week6/forcomp/src/main/scala/forcomp/Anagrams.scala [hunk 3]
```diff
@@ -92,7 +109,16 @@ object Anagrams extends AnagramsInterface {
    *  Note: the resulting value is an occurrence - meaning it is sorted
    *  and has no zero-entries.
    */
-  def subtract(x: Occurrences, y: Occurrences): Occurrences = ???
+  def subtract(x: Occurrences, y: Occurrences): Occurrences = {
+    val yMap = y.toMap.withDefaultValue(0)
+    for {
+      (xc,xi) <- x
```
Comment (maryum375) [inline Scala2-Functional-Programming/Week6/forcomp/src/main/scala/forcomp/Anagrams.scala:115]: Lost you here with those variable names...
Please change

File: Scala2-Functional-Programming/Week6/forcomp/src/main/scala/forcomp/Anagrams.scala [hunk 4]
```diff
@@ -134,7 +160,26 @@ object Anagrams extends AnagramsInterface {
    *
    *  Note: There is only one anagram of an empty sentence.
    */
-  def sentenceAnagrams(sentence: Sentence): List[Sentence] = ???
+  def sentenceAnagrams(sentence: Sentence): List[Sentence] = {
+    val lettersOcc: Occurrences = sentenceOccurrences(sentence)
+
+    def helper(current: Sentence, leftOccurrence: Occurrences): List[Sentence]  = {
```
Comment (maryum375) [inline Scala2-Functional-Programming/Week6/forcomp/src/main/scala/forcomp/Anagrams.scala:166]: Not a big fan of function named `helper`. 
Those names are usually given to functions that help with more than one thing.
Please rename or make sure that function really does one single thing

Comment (maryum375) [inline Scala2-Functional-Programming/Week6/forcomp/src/main/scala/forcomp/Anagrams.scala:166]: There is a simpler way to implement this, a way where this function accepts only one parameter.
Choose whether you want to invest time to implement it this way.
I encourage you to do so.

File: Scala2-Functional-Programming/Week6/forcomp/src/main/scala/forcomp/Anagrams.scala [hunk 5]
```diff
@@ -145,8 +190,8 @@ object Dictionary {
       sys.error("Could not load word list, dictionary file not found")
     }
     try {
-      val s = scala.io.Source.fromInputStream(wordstream)
-      s.getLines.toList
+      val s = Source.fromInputStream(wordstream)(Codec.UTF8)
```
Comment (maryum375) [inline Scala2-Functional-Programming/Week6/forcomp/src/main/scala/forcomp/Anagrams.scala:193]: Why explicitly adding the Codec?

File: Scala2-Functional-Programming/Week6/forcomp/src/main/scala/forcomp/Anagrams.scala [hunk 6]
```diff
@@ -145,8 +190,8 @@ object Dictionary {
...
+      s.getLines().toList
```
Comment (maryum375) [inline Scala2-Functional-Programming/Week6/forcomp/src/main/scala/forcomp/Anagrams.scala:194]: As you probably understood, the `()` is not really needed

**Review by maryum375 — CHANGES_REQUESTED** [review]

File: Scala2-Functional-Programming/Week6/forcomp/src/main/scala/forcomp/Anagrams.scala (hunk 6 above)
Comment (taljmars) [reply Scala2-Functional-Programming/Week6/forcomp/src/main/scala/forcomp/Anagrams.scala:194]: I think this change was done by the IDE..

**Review by maryum375 — COMMENTED** [review]
Good job :)

File: Scala2-Functional-Programming/Week6/forcomp/src/main/scala/forcomp/Anagrams.scala (hunk 2 above)
Comment (taljmars) [reply Scala2-Functional-Programming/Week6/forcomp/src/main/scala/forcomp/Anagrams.scala]: Done

File: Scala2-Functional-Programming/Week6/forcomp/src/main/scala/forcomp/Anagrams.scala (hunk 1 above)
Comment (taljmars) [reply Scala2-Functional-Programming/Week6/forcomp/src/main/scala/forcomp/Anagrams.scala]: Done

File: Scala2-Functional-Programming/Week6/forcomp/src/main/scala/forcomp/Anagrams.scala (hunk 3 above)
Comment (taljmars) [reply Scala2-Functional-Programming/Week6/forcomp/src/main/scala/forcomp/Anagrams.scala]: Done

File: Scala2-Functional-Programming/Week6/forcomp/src/main/scala/forcomp/Anagrams.scala (hunk 4 above)
Comment (taljmars) [reply Scala2-Functional-Programming/Week6/forcomp/src/main/scala/forcomp/Anagrams.scala]: change the name, I would like to move on to other items, if you have a different solution I will gladly see it

**Review by maryum375 — APPROVED** [review]

**Pull request merged**